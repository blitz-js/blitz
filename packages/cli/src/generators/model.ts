import Generator, {GeneratorOptions} from '../generator'
import enquirer from 'enquirer'
import path from 'path'
import {log} from '@blitzjs/server'

export interface AppGeneratorOptions extends GeneratorOptions {
  name: string
  pluralName: string
}

class ModelGenerator extends Generator<AppGeneratorOptions> {
  // default subdirectory is /app/[name], we need to back out of there to generate the model
  static subdirectory = '../..'
  static template = ''

  async write() {
    try {
      if (!this.fs.exists(path.resolve('db/schema.prisma'))) {
        throw new Error('Prisma schema file was not found')
      }
      const {fields: fieldsString} = await enquirer.prompt({
        name: 'fields',
        type: 'input',
        message: `Please input all fields which should exist on a ${this.options.name} as a space-separated list of key:value pairs.
An 'id' column will be auto-generated and should not be entered.
Modifiers can be denoted with a trailing '?' (e.g. 'lastname:string?' or 'users:User[]')`,
      })
      const fields = fieldsString.split(' ')
      let modelDefinition = `\nmodel ${this.options.name} {
  id  Int  @id @default(autoincrement())`
      for (const field of fields) {
        const [fieldName, fieldType] = field.split(':')
        modelDefinition += `\n  ${fieldName}  ${fieldType}`
      }
      modelDefinition += '\n}\n'
      this.fs.append(path.resolve('db/schema.prisma'), modelDefinition)
    } catch (error) {
      throw error
    }
  }

  async postWrite() {
    log.success(`Successfully created prisma model for ${this.options.name.toLocaleLowerCase()}`)
  }
}

export default ModelGenerator
