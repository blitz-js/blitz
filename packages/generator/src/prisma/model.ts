import {Field, FieldType} from "./field"
import {singlePascal} from "../utils/plurals"

function stringifyFieldsForPrinting(fields: Field[]) {
  let maxNameLength = 1
  let maxTypeLength = 1
  for (let field of fields) {
    const [name, type] = field.toString().split(/[\s]+/)
    maxNameLength = Math.max(maxNameLength, name.length)
    maxTypeLength = Math.max(maxTypeLength, type.length)
  }
  // add horizontal padding, otherwise there would be no space after the type
  maxNameLength += 2
  maxTypeLength += 2
  return fields.map((field) => {
    const [name, type, ...rest] = field.toString().split(/[\s]+/)
    const attrs = rest.join(" ")
    const namePad = maxNameLength - name.length
    const typePad = maxTypeLength - type.length
    return `${name}${Array(namePad).join(" ")}${type}${Array(typePad).join(" ")}${attrs}`
  })
}

export class Model {
  name: string
  fields: Field[]

  constructor(name: string, fields: Field[] = []) {
    this.name = singlePascal(name)
    this.fields = fields
  }

  private getIdField() {
    return new Field("id", {
      isRequired: true,
      isList: false,
      isId: true,
      default: "autoincrement()",
      type: FieldType.Int,
    })
  }

  private getCreatedAtField() {
    return new Field("createdAt", {
      isRequired: true,
      isList: false,
      isId: false,
      default: "now()",
      type: FieldType.DateTime,
    })
  }

  private getUpdatedAtField() {
    return new Field("updatedAt", {
      isRequired: true,
      isList: false,
      isId: false,
      isUpdatedAt: true,
      type: FieldType.DateTime,
    })
  }

  private getFields() {
    return stringifyFieldsForPrinting([
      this.getIdField(),
      this.getCreatedAtField(),
      this.getUpdatedAtField(),
      ...this.fields,
    ])
      .map((field) => `\n  ${field}`)
      .join("")
  }

  toString() {
    return `model ${this.name} {${this.getFields()}\n}`
  }
}
