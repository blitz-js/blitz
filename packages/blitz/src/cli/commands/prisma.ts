import {CliCommand} from "../index"
import arg from "arg"
import {runPrisma} from "../../utils/run-prisma"

export const runPrismaExitOnError = async (...args: Parameters<typeof runPrisma>) => {
  const result = await runPrisma(...args)

  if (!result.success) {
    process.exit(1)
  }
}

const prisma: CliCommand = async () => {
  const args = arg(
    {},
    {
      permissive: true,
    },
  )

  await runPrismaExitOnError(args._.slice(1))
}

export {prisma}
