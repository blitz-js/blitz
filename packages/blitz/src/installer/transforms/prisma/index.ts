export * from "./add-prisma-enum"
export * from "./add-prisma-field"
export * from "./add-prisma-generator"
export * from "./add-prisma-model-attribute"
export * from "./add-prisma-model"
export * from "./produce-schema"
export * from "./set-prisma-data-source"
