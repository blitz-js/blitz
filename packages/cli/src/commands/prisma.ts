import {Command, flags} from "@oclif/command"

const getPrismaBin = () => require("@blitzjs/server").resolveBinAsync("@prisma/cli", "prisma")
let prismaBin: string

export const runPrisma = async (args: string[], silent = false) => {
  if (!prismaBin) {
    try {
      prismaBin = await getPrismaBin()
    } catch (err) {
      throw err
    }
  }

  const cp = require("cross-spawn").spawn(prismaBin, args, {
    stdio: silent ? "ignore" : "inherit",
    env: process.env,
  })
  const code = await require("p-event")(cp, "exit", {rejectionEvents: []})

  return code === 0
}

export const runPrismaExitOnError = async (...args: Parameters<typeof runPrisma>) => {
  const success = await runPrisma(...args)

  if (!success) {
    process.exit(1)
  }
}

export class PrismaCommand extends Command {
  static description = "Loads env variables then proxies all args to Prisma CLI"
  static aliases = ["p"]

  static flags = {
    help: flags.help({char: "h"}),
  }

  static strict = false

  async run() {
    const {argv} = this.parse(PrismaCommand)

    await runPrismaExitOnError(argv)
  }
}
