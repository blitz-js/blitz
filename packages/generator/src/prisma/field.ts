import {singlePascal, singleCamel, pluralCamel} from "../utils/plurals"
import {log} from "@blitzjs/display"

export enum FieldType {
  Boolean = "Boolean",
  DateTime = "DateTime",
  Float = "Float",
  Int = "Int",
  Json = "Json",
  String = "String",
}

export enum Relation {
  hasOne,
  hasMany,
  belongsTo,
}

function isRelation(maybeRelation: string): maybeRelation is keyof typeof Relation {
  return Object.keys(Relation).includes(maybeRelation)
}

interface FieldArgs {
  default?: any
  isId?: boolean
  isList?: boolean
  isRequired?: boolean
  isUnique?: boolean
  isUpdatedAt?: boolean
  type: FieldType | string
  relationFromFields?: string[]
  relationToFields?: string[]
}

const fallbackIfUndef = <T extends any>(defaultValue: T, input?: T) => {
  if (input === undefined) return defaultValue
  return input
}

const defaultValueTest = /\[([\w]+)\]/
const builtInGenerators = ["autoincrement", "now", "uuid", "cuid"]

class MissingFieldNameError extends Error {}

export class Field {
  default?: any
  name: string
  isId: boolean
  isList: boolean
  isRequired: boolean
  isUnique: boolean
  isUpdatedAt: boolean
  type: FieldType | string
  relationFromFields?: string[]
  relationToFields?: string[]

  // 'name:type?[]:attribute' => Field
  static parse(input: string): Field[] {
    const [_fieldName, _fieldType = "String", attribute] = input.split(":")
    let fieldName = singleCamel(_fieldName)
    let fieldType = singlePascal(_fieldType)
    let isRequired = true
    let isList = false
    let isUpdatedAt = false
    let isUnique = false
    let defaultValue = undefined
    let relationFromFields = undefined
    let relationToFields = undefined
    let maybeIdField = undefined
    if (fieldType.includes("?")) {
      fieldType = fieldType.replace("?", "")
      isRequired = false
    }
    if (fieldType.includes("[]")) {
      fieldType = fieldType.replace("[]", "")
      fieldName = pluralCamel(fieldName)
      isList = true
    }
    // use original unmodified field name in case the list handling code
    // has modified fieldName
    if (isRelation(_fieldName)) {
      // this field is an object type, not a scalar type
      const relationType = Relation[_fieldName]
      // translate the type into the name since they should stay in sync
      fieldName = singleCamel(fieldType)
      fieldType = singlePascal(fieldType)

      switch (relationType) {
        case Relation.hasOne:
          // current model gets single `modelName ModelName` association field
          isList = false
          break
        case Relation.hasMany:
          // current model gets single `modelNames ModelName[]` association field
          fieldName = pluralCamel(fieldName)
          isList = true
          isRequired = true
          break
        case Relation.belongsTo:
          // current model gets two fields:
          //   modelName    ModelName   @relation(fields: [modelNameId], references: [id])
          //   modelNameId  ModelIdType
          const idFieldName = `${fieldName}Id`
          relationFromFields = [idFieldName]
          relationToFields = ["id"]
          maybeIdField = new Field(idFieldName, {type: FieldType.Int, isRequired})
          isList = false
          break
      }
    }
    if (!/^[A-Za-z]*$/.test(fieldName)) {
      // modelName should be just alpha characters at this point, validate
      throw new Error(
        `[Field.parse]: received unknown special character in field name: ${fieldName}`,
      )
    }

    if (/unique/i.test(attribute)) isUnique = true
    if (/updatedAt/i.test(attribute)) isUpdatedAt = true
    if (/default/i.test(attribute)) {
      if (defaultValueTest.test(attribute)) {
        const [, _defaultValue] = attribute.match(defaultValueTest)!
        defaultValue = builtInGenerators.includes(_defaultValue)
          ? {name: _defaultValue}
          : _defaultValue
      }
    }
    try {
      const parseResult = new Field(fieldName, {
        default: defaultValue,
        isId: false,
        isList,
        isRequired,
        isUnique,
        isUpdatedAt,
        relationFromFields,
        relationToFields,
        type: fieldType,
      })
      return maybeIdField ? [parseResult, maybeIdField] : [parseResult]
    } catch (err) {
      if (err instanceof MissingFieldNameError) {
        throw new Error(
          `Each field in a model must have a name, but you supplied ${log.variable(
            input,
          )}.\n         Try giving the field as a ${log.variable(
            "name:type",
          )} pair, such as ${log.variable("description:string")}.`,
        )
      }
      throw err
    }
  }

  constructor(name: string, options: FieldArgs) {
    if (!name) throw new MissingFieldNameError("[PrismaField]: A field name is required")
    if (!options.type) {
      log.warning(
        `No field type specified for field ${log.variable(name)}, falling back to ${log.variable(
          "String",
        )}.`,
      )
      options.type = FieldType.String
    }
    this.name = name
    this.isList = fallbackIfUndef(false, options.isList)
    this.isRequired = fallbackIfUndef(true, options.isRequired)
    this.isUnique = fallbackIfUndef(false, options.isUnique)
    this.isId = fallbackIfUndef(false, options.isId)
    this.type = options.type
    this.default = options.default
    this.isUpdatedAt = fallbackIfUndef(false, options.isUpdatedAt)
    this.relationFromFields = options.relationFromFields
    this.relationToFields = options.relationToFields
    if (!this.isRequired && this.isList) {
      throw new Error("[PrismaField]: a type cannot be both optional and a list")
    }
    if (this.isId && this.default === undefined) {
      throw new Error("[PrismaField]: ID fields must have a default value")
    }
  }

  private getDefault() {
    if (this.default === undefined) return ""
    let defaultValue: string
    if (typeof this.default === "object") {
      // { name: 'fnname' } is based off of the Prisma model definition
      defaultValue = `${this.default.name}()`
    } else {
      defaultValue = String(this.default)
    }
    return `@default(${defaultValue})`
  }

  private getId() {
    return this.isId ? "@id" : ""
  }

  private getIsUnique() {
    return this.isUnique ? "@unique" : ""
  }

  private getIsUpdatedAt() {
    return this.isUpdatedAt ? "@updatedAt" : ""
  }

  private getRelation() {
    if (this.relationFromFields === undefined || this.relationToFields === undefined) return ""
    const separator =
      this.relationToFields &&
      this.relationToFields.length > 0 &&
      this.relationFromFields &&
      this.relationFromFields.length
        ? ", "
        : ""
    const fromFields =
      this.relationFromFields && this.relationFromFields.length > 0
        ? `fields: [${this.relationFromFields.toString()}]`
        : ""

    const toFields =
      this.relationToFields && this.relationToFields.length > 0
        ? `references: [${this.relationToFields.toString()}]`
        : ""
    return `@relation(${fromFields}${separator}${toFields})`
  }

  private getTypeModifiers() {
    return `${this.isRequired ? "" : "?"}${this.isList ? "[]" : ""}`
  }

  private getAttributes() {
    const possibleAttributes = [
      this.getDefault(),
      this.getId(),
      this.getIsUnique(),
      this.getIsUpdatedAt(),
      this.getRelation(),
    ]
    // filter out any attributes that return ''
    const attrs = possibleAttributes.filter((attr) => attr)
    if (attrs.length > 0) {
      return `  ${attrs.join(" ")}`
    }
    return ""
  }

  toString() {
    return `${this.name}  ${this.type}${this.getTypeModifiers()}${this.getAttributes()}`
  }
}
