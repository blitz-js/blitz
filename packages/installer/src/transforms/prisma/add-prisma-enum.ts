import {Enum} from "@mrleebo/prisma-ast"
import {produceSchema} from "./produce-schema"

/**
 * Adds an enum to your schema.prisma data model.
 *
 * @param source - schema.prisma source file contents
 * @param enumProps - the enum to add
 * @returns The modified schema.prisma source
 * @example Usage
 * ```
 *  addPrismaEnum(source, {
      type: "enum",
      name: "Role",
      enumerators: [
        {type: "enumerator", name: "USER"},
        {type: "enumerator", name: "ADMIN"},
      ],
    })
 * ```
 */
export function addPrismaEnum(source: string, enumProps: Enum): Promise<string> {
  return produceSchema(source, (schema) => {
    const existing = schema.list.find((x) => x.type === "enum" && x.name === enumProps.name)
    existing ? Object.assign(existing, enumProps) : schema.list.push(enumProps)
  })
}
