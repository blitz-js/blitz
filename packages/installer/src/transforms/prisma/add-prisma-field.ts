import {Field, Model} from "@mrleebo/prisma-ast"
import {produceSchema} from "./produce-schema"

/**
 * Adds a field to a model in your schema.prisma data model.
 *
 * @param source - schema.prisma source file contents
 * @param modelName - name of the model to add a field to
 * @param fieldProps - the field to add
 * @returns The modified schema.prisma source
 * @example Usage
 * ```
 *  addPrismaField(source, "Project", {
      type: "field",
      name: "name",
      fieldType: "String",
      optional: false,
      attributes: [{type: "attribute", kind: "field", name: "unique"}],
    })
 * ```
 */
export function addPrismaField(
  source: string,
  modelName: string,
  fieldProps: Field,
): Promise<string> {
  return produceSchema(source, (schema) => {
    const model = schema.list.find((x) => x.type === "model" && x.name === modelName) as Model
    if (!model) return

    const existing = model.properties.find((x) => x.type === "field" && x.name === fieldProps.name)
    existing ? Object.assign(existing, fieldProps) : model.properties.push(fieldProps)
  })
}
