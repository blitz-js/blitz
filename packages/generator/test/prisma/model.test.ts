import {Field} from "../../src/prisma/field"
import {Model} from "../../src/prisma/model"

describe("Prisma Model", () => {
  it("generates a proper model", () => {
    expect(
      new Model(
        "user",
        [
          Field.parse("email:string:unique"),
          Field.parse("updated:dateTime:updatedAt"),
          Field.parse("recentLogins:dateTime[]"),
          Field.parse("twoFactorEnabled:boolean"),
          Field.parse("twoFactorMethod:string?"),
        ].flat(),
      ).toString(),
    ).toMatchSnapshot()
  })
})
