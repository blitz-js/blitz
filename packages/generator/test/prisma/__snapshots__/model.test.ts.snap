// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prisma Model generates a proper model 1`] = `
"
model User {
  id               Int        @id @default(autoincrement())
  createdAt        DateTime   @default(now())
  updatedAt        DateTime   @updatedAt
  email            String     @unique
  updated          DateTime   @updatedAt
  recentLogins     DateTime[]
  twoFactorEnabled Boolean
  twoFactorMethod  String?
}
"
`;
